pkg/receive/handler.go

```diff
@@ -629,6 +599,77 @@ func isNotReady(err error) bool {
 		status.Code(err) == codes.Unavailable
 }
 
+// countCause counts the number of errors within the given error slice
+// whose causes satisfy the given function.
+// cause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
+func countCause(errs []error, f func(error) bool) int {
+	var n int
+	for _, err := range errs {
+		if f(errors.Cause(err)) {
+			n++
+		}
+	}
+	return n
+}
+
+type expectedErrors []*struct {
+	err   error
+	cause func(error) bool
+	count int
+}
+
+func (a expectedErrors) Len() int           { return len(a) }
+func (a expectedErrors) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a expectedErrors) Less(i, j int) bool { return a[i].count < a[j].count }
+
+// cause returns a known error which has occurred more than the given threshold.
+// cause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
+func cause(err error, threshold int) error {
+	if err == nil {
+		return nil
+	}
+	unwrappedErr := errors.Cause(err)
+	errs, ok := unwrappedErr.(terrors.MultiError)
+	if !ok {
+		errs = []error{unwrappedErr}
+	}
+	if threshold > 0 {
+		expErrs := expectedErrors{
```

Comment:
You know that for those two error types, we can just remove `countCause` and implement all directly here and it will use less lines and lower complexity? (: Or am I missing something?